lib/compress/zstdmt_compress.c

```diff
@@ -1818,6 +1835,60 @@ static int ZSTDMT_tryGetInputRange(ZSTDMT_CCtx* mtctx)
     return 1;
 }
 
+/**
+ * Searches through the input for a synchronization point. If one is found, we
+ * will instruct the caller to flush, and return the number of bytes to load.
+ * Otherwise, we will load as many bytes as possible and continue as normal.
+ */
+static size_t findSynchronizationPoint(ZSTDMT_CCtx* mtctx, ZSTD_inBuffer* input, ZSTD_EndDirective* endOp) {
+    size_t toLoad = MIN(input->size - input->pos, mtctx->targetSectionSize - mtctx->inBuff.filled);
+    if (!mtctx->params.rsyncable)
+        /* Rsync is disabled. */
+        return toLoad;
+    if (mtctx->inBuff.filled + toLoad < RSYNC_LENGTH)
+      /* Not enough to compute the hash.
+       * We will miss any synchronization points in this RSYNC_LENGTH byte
+       * window. However, since it depends only in the internal buffers, if the
+       * state is already synchronized, we will remain synchronized.
+       * Additionally, the probability that we miss a synchronization point is
+       * low: RSYNC_LENGTH / targetSectionSize.
+       */
+      return toLoad;
+    if (mtctx->inBuff.filled < RSYNC_LENGTH) {
+      /* Prefill the buffer to be at least RSYNC_LENGTH. */
+      size_t const toPrefill = MIN(toLoad, RSYNC_LENGTH - mtctx->inBuff.filled);
+      memcpy((char*)mtctx->inBuff.buffer.start + mtctx->inBuff.filled, (const char*)input->src + input->pos, toPrefill);
+      mtctx->inBuff.filled += toPrefill;
+      input->pos += toPrefill;
+      toLoad -= toPrefill;
+    }
+    assert(mtctx->inBuff.filled >= RSYNC_LENGTH);
+    {   /* Starting with the hash of the previous RSYNC_LENGTH bytes, roll
+         * through the input. If we hit a synchronization point, then cut the
+         * job off, and tell the compressor to flush the job. Otherwise, load
+         * all the bytes and continue as normal.
+         * If we go too long without a synchronization point (targetSectionSize)
+         * then a block will be emitted anyways, but this is okay, since if we
+         * are already synchronized we will remain synchronized.
+         */
+        U64 const primePower = mtctx->rsync.primePower;
+        U64 const hitMask = mtctx->rsync.hitMask;
+        BYTE const* const start = (BYTE const*)input->src + input->pos;
+        BYTE const* prev = (BYTE const*)mtctx->inBuff.buffer.start + mtctx->inBuff.filled - RSYNC_LENGTH;
+        U64 hash = ZSTD_rollingHash_compute(prev, RSYNC_LENGTH);
+        size_t pos;
+        for (pos = 0; pos < toLoad; ++pos) {
+            BYTE const toRemove = pos < RSYNC_LENGTH ? prev[pos] : start[pos - RSYNC_LENGTH];
+            hash = ZSTD_rollingHash_rotate(hash, toRemove, start[pos], primePower);
+            if ((hash & hitMask) == hitMask) {
+                if (*endOp == ZSTD_e_continue)
+                    *endOp = ZSTD_e_flush;
+                return pos + 1;
```

Comment:
My understanding is that this function returns the nb of bytes to read from input and to pass to create next job.
It also provides the associated `ZSTD_EndDirective`.

Generally speaking, this function has multiple side effects to pay attention to : the return value, `input->pos` and `*endOp`, that may or may not have been modified, `mtctx` itself can be modified. 

It's meant to blend into existing code, and seems to do so very well. 

However, when decoded in isolation, the function has a strange set of combined actions,
which makes it difficult to re-use it, in any other context than the single one it's supposed to blend into.

We could maybe let it this way, and just list all its side effects, to make sure that any future programmer trying to re-use or modify this function doesn't miss any of them.

A more difficult alternative could be to make this function a bit more "clean" : do not modify `mtctx`, to not modify `input`, do not modify `endOp`, provide both results `toLoad` and `endOp` into return value (which must be a `struct` then). I understand it's more complex and requires an internal buffer to cope with situations when `inBuff.filled < RSYNC_LENGTH`. And since this function is not intended to be re-used anywhere else, it might be overkill.

So feel free to select one or the other solutions.